sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala

```diff
@@ -432,113 +789,137 @@ object GpuWindowExec extends Arm {
-  def computeRunning(
-      iter: Iterator[ColumnarBatch],
-      boundWindowOps: Seq[GpuExpression],
-      boundPartitionSpec: Seq[Expression],
-      numOutputBatches: GpuMetric,
-      numOutputRows: GpuMetric,
-      opTime: GpuMetric): Iterator[ColumnarBatch] = {
-    var lastParts: Array[Scalar] = Array.empty
-    val fixers = fixerIndexMap(boundWindowOps)
-
-    def saveLastParts(newLastParts: Array[Scalar]): Unit = {
-      lastParts.foreach(_.close())
-      lastParts = newLastParts
-    }
+/**
+ * An iterator that can do row based aggregations on running window queries (Unbounded preceding to
+ * current row) if and only if the aggregations are instances of GpuBatchedRunningWindowFunction
+ * which can fix up the window output when an aggregation is only partly done in one batch of data.
+ * Because of this there is no requirement about how the input data is batched, but it  must
+ * be sorted by both partitioning and ordering.
+ */
+class GpuRunningWindowIterator(
+    input: Iterator[ColumnarBatch],
+    override val boundWindowOps: Seq[GpuExpression],
+    override val boundPartitionSpec: Seq[GpuExpression],
+    override val boundOrderSpec: Seq[SortOrder],
+    numOutputBatches: GpuMetric,
+    numOutputRows: GpuMetric,
+    opTime: GpuMetric) extends Iterator[ColumnarBatch] with BasicWindowCalc {
+
+  import GpuRunningWindowIterator._
+
+  private var internalFixerIndexMap: Option[Map[Int, BatchedRunningWindowFixer]] = None
+  // This should only ever be cached in between calls to has next and next. This is just
+  // to let us filter out empty batches.
+  private var cachedBatch: Option[ColumnarBatch] = None
+  private var lastParts: Array[Scalar] = Array.empty
+
+  private def saveLastParts(newLastParts: Array[Scalar]): Unit = {
+    lastParts.foreach(_.close())
+    lastParts = newLastParts
+  }
 
-    def closeState(): Unit = {
-      saveLastParts(Array.empty)
-      fixers.values.foreach(_.close())
+  def close(): Unit = {
+    internalFixerIndexMap.foreach { m =>
+      m.values.foreach(_.close())
     }
-
-    TaskContext.get().addTaskCompletionListener[Unit](_ => closeState())
-
-    iter.map { cb =>
-      val numRows = cb.numRows
-      numOutputBatches += 1
-      numOutputRows += numRows
-      withResource(new MetricRange(opTime)) { _ =>
-        val fullWindowProjectList = boundWindowOps ++ boundPartitionSpec
-        withResource(
-          GpuProjectExec.projectAndClose(cb, fullWindowProjectList, NoopMetric)) { full =>
-          // part columns are owned by full and do not need to be closed, but should not be used
-          // if full is closed
-          val partColumns = boundPartitionSpec.indices.map { idx =>
-            full.column(idx + boundWindowOps.length).asInstanceOf[GpuColumnVector].getBase
+    internalFixerIndexMap = None
+    saveLastParts(Array.empty)
+  }
+  TaskContext.get().addTaskCompletionListener[Unit](_ => close())
+
+  private def fixerIndexMap: Map[Int, BatchedRunningWindowFixer] = {
```

Comment:
It is because I wanted close to be idempotent. If I make it a lazy val then I will need a separate var to track that I have closed it. Which I can do if you think it is going to be cleaner.

Reply:
OK it is cleaner I'll do it your way

Reply:
> It is because I wanted close to be idempotent.

Ah, right.  One small issue then with a `lazy val` is that even with an `isClosed` variable we could end up instantiating the lazy val just to turn around and close it during the `close()` method if we never ended up touching the value before `close()` is called.  I assume this isn't a concern since we would expect the lazy value to be touched in almost all cases before the instance is closed.

Times like this I wish it was simple to check if a lazy value has been initialized.  Can be done manually but it's not pretty.